_posts/2019/2019-05-19-dont-email-from-activerecord-callbacks.md

```diff
@@ -0,0 +1,71 @@
+---
+title: "Don’t Email From Active Record Callbacks"
+description: "It’s in the docs, but it’s a bad idea."
+layout: article
+category: ruby
+image:
+  base: '2019/dont-email-from-activerecord-callbacks'
+  alt: 'Envelopes'
+  credit: Joanna Kosinska
+  source: "https://unsplash.com/photos/uGcDWKN91Fs"
+
+---
+
+Among the first things you often want to do with your Rails application is send email.
+
+A frequent pattern is that an email is often sent after a change to, or creation of, one of your models.
+
+
+## Instead of…
+
+…sending email inside a callback in your model.
+
+```ruby
+class Comment < ApplicationRecord
+  after_create :send_email_to_author
+
+  private
+
+  def send_email_to_author
+    AuthorMailer.
+      with(author: author).
+      comment_notification.
+      deliver_now
+  end
+end
+```
+
+
+## Use…
+
+…the controller to send your email separately from the model.
+
+```ruby
+class CommentsController < ApplicationController
+  def create
+    Comment.create(comment_params)
+    AuthorMailer.
+      with(author: author).
+      comment_notification.
+      deliver_now
+  end
+end
+```
+
+
+## Why?
+
+This boils down, for me, to clarity and preventing unintended side effects.
+
+Invariably at some point you’ll need to create a comment without emailing the author; perhaps in the `rails console` or in another action. You probably don’t want to send email in those cases. You _can_ use methods that [skip callbacks](https://guides.rubyonrails.org/active_record_callbacks.html#skipping-callbacks), but then you’re into a whole new world of complexity.
+
+I find it clearer to have a simple, procedural, 'list of things to do' inside my controller actions. I also find that debugging side-effects through a model’s callbacks to be much more cognitively difficult.
```

Comment:
```suggestion
It's clearer to have a simple, procedural, 'list of things to do' inside my controller actions. I also find that debugging side-effects through a model’s callbacks are much more cognitively difficult.
```

Reply:
I think you need to explain _why_ for both of the things you mention here: why is it clearer to have the list in the controller? (Because you can read from A to Z?) And why is it more cognitively difficult to debug side-effects from callbacks? (Because you have to keep more context in your head as to what's happening when?)